docs/gguf.md

```diff
@@ -0,0 +1,546 @@
+    GGML_TYPE_Q5_K = 13,
+    GGML_TYPE_Q6_K = 14,
+    GGML_TYPE_Q8_K = 15,
+    GGML_TYPE_I8,
+    GGML_TYPE_I16,
+    GGML_TYPE_I32,
+    GGML_TYPE_COUNT,
+};
+
+enum gguf_metadata_value_type: uint32_t {
+    // The value is a 8-bit unsigned integer.
+    GGUF_METADATA_VALUE_TYPE_UINT8 = 0,
+    // The value is a 8-bit signed integer.
+    GGUF_METADATA_VALUE_TYPE_INT8 = 1,
+    // The value is a 16-bit unsigned little-endian integer.
+    GGUF_METADATA_VALUE_TYPE_UINT16 = 2,
+    // The value is a 16-bit signed little-endian integer.
+    GGUF_METADATA_VALUE_TYPE_INT16 = 3,
+    // The value is a 32-bit unsigned little-endian integer.
+    GGUF_METADATA_VALUE_TYPE_UINT32 = 4,
+    // The value is a 32-bit signed little-endian integer.
+    GGUF_METADATA_VALUE_TYPE_INT32 = 5,
+    // The value is a 32-bit IEEE754 floating point number.
+    GGUF_METADATA_VALUE_TYPE_FLOAT32 = 6,
+    // The value is a boolean.
+    // 1-byte value where 0 is false and 1 is true.
+    // Anything else is invalid, and should be treated as either the model being invalid or the reader being buggy.
+    GGUF_METADATA_VALUE_TYPE_BOOL = 7,
+    // The value is a UTF-8 non-null-terminated string, with length prepended.
+    GGUF_METADATA_VALUE_TYPE_STRING = 8,
+    // The value is an array of other values, with the length and type prepended.
+    ///
+    // Arrays can be nested, and the length of the array is the number of elements in the array, not the number of bytes.
+    GGUF_METADATA_VALUE_TYPE_ARRAY = 9,
+}
+
+// A string in GGUF.
+struct gguf_string_t {
+    // The length of the string, in bytes.
+    uint32_t len;
+    // The string as a UTF-8 non-null-terminated string.
+    char string[len];
+}
+
+union gguf_metadata_value_t {
+    uint8_t uint8;
+    int8_t int8;
+    uint16_t uint16;
+    int16_t int16;
+    uint32_t uint32;
+    int32_t int32;
+    float float32;
+    bool bool_;
+    gguf_string_t string;
+    struct {
+        // Number of elements, not bytes
+        uint32_t len;
+        // Any value type is valid, including arrays.
+        gguf_metadata_value_type type;
+        // The array of values.
+        gguf_metadata_value_t array[len];
+    } array;
+};
+
+struct gguf_metadata_kv_t {
+    // The key of the metadata. It is a standard GGUF string, with the following caveats:
+    // - It must be a valid ASCII string.
+    // - It must be a hierarchical key, where each segment is `lower_snake_case` and separated by a `.`.
+    // - It must be at most 2^16-1/65535 bytes long.
+    // Any keys that do not follow these rules are invalid.
+    gguf_string_t key;
+
+    // The length of the value, in bytes
+    uint32_t value_len;
+    // The type of the value.
+    // Must be one of the `gguf_metadata_value_type` values.
+    gguf_metadata_value_type value_type;
+    // The value.
+    gguf_metadata_value_t value;
+};
+
+struct gguf_header_t {
+    // Magic number to announce that this is a GGUF file.
+    // Must be `'GGUF'`/`0x47475546`.
```

Comment:
This should by reversed so it is written as `GGUF` in the model file. `0x46554747`

Reply:
It's written as 'GGUF' if you look at it bytewise (0x47 0x47 0x55 0x46 is G G U F), but not if you look at it as a 32-bit little-endian integer. I think it's better to keep it this way to ensure that little/big-endian isn't a concern?

Reply:
In https://github.com/ggerganov/llama.cpp/pull/2398 we already changed this. Bu it could possibly be reversed. @ggerganov

Reply:
...too late it seems, llama.cpp's implementation seems to do little-endian. OK.

Reply:
I've updated this, but I think it makes more sense the way it was specified before (i.e. as a 4-byte string, not as a little-endian 4-byte integer).

Reply:
...okay, turns out it actually _is_ `GGUF` in the resulting file right now, it's just the little endian cancels out:
```
# hexdump -C models/llama2/llama-2-7b-f16.gguf | head -n 1 
00000000  47 47 55 46 01 00 00 00  23 01 00 00 0c 00 00 00  |GGUF....#.......|
```

and this is because the header is written as a little-endian integer:
`self.fout.write(struct.pack("<I", GGUF_MAGIC))`

but the constant is *already* little-endian:
`GGUF_MAGIC             = 0x46554747`

so in the write it gets reversed and becomes 'GGUF'. I'm going to mention that's what it is in the spec, but both the Python and C/C++ should probably be updated to not treat them as integers O_o

Reply:
Using 0x46554747 works in C++
But yes it would probably be less confusing to write the magic as 4 bytes instead.